### PR TITLE
Issue #2502497 by Pol: Use Drupal 7 Legacy service everywhere

### PR DESCRIPTION
**Pull request by drupol**
Now that we have a legacy Drupal 7 service, we can use it when using Drupal function in other service(s).

### CODE REVIEW

File: src/Extension/ModuleHandler.php [hunk 1]
```diff
@@ -41,7 +49,8 @@ class ModuleHandler implements ModuleHandlerInterface {
    * @see \Drupal\Core\DrupalKernel
    * @see \Drupal\Core\CoreServiceProvider
    */
-  public function __construct($root, array $module_list = array()) {
+  public function __construct(Drupal7 $drupal7, $root, array $module_list = array()) {
```
Comment (fabianx-ai) [inline src/Extension/ModuleHandler.php]: Lets put drupal7 2nd, also need to update the doxygen, which is wrong anyway.

Scrutinizer reported that something calls this statically (createInstance or so) and misses the new argument.

File: src/Logger/WatchdogLogger.php [hunk 2]
```diff
@@ -18,7 +19,22 @@
...
 
   /**
+   * The Drupal7 service.
+   *
+   * @var \Drupal\service_container\Legacy\Drupal7
+   */
+  protected $drupal7;
+
+  /**
+   * {@inheritdoc}
```
Comment (fabianx-ai) [inline src/Logger/WatchdogLogger.php]: Can't be defined, must fill out.

File: src/Logger/WatchdogLogger.php [hunk 3]
```diff
@@ -18,7 +19,22 @@
...
+   */
+  function __construct(Drupal7 $drupal7) {
+    $this->drupal7 = $drupal7;
+  }
+
+  /**
    * {@inheritdoc}
+   * @codeCoverageIgnore
```
Comment (fabianx-ai) [inline src/Logger/WatchdogLogger.php]: Not needed, we want to remove @codeCoverageIgnore not add new ones ...

File: src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php [hunk 4]
```diff
@@ -41,11 +41,12 @@ public function getContainerDefinition() {
 
     $services['module_handler'] = array(
       'class' => '\Drupal\service_container\Extension\ModuleHandler',
-      'arguments' => array(DRUPAL_ROOT),
+      'arguments' => array('@drupal7', DRUPAL_ROOT),
```
Comment (fabianx-ai) [inline src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php]: Lets reverse arguments.

File: src/Logger/WatchdogLogger.php (hunk 2 above)
Comment (drupol) [inline src/Logger/WatchdogLogger.php]: This is because we are using the watchdog function from D7 in the ::log() method. Should I not use it ?

Comment (fabianx-ai) [inline src/Logger/WatchdogLogger.php]: I mean you need to add the doxygen:
-   \* @param \Drupal\service_container\Legacy\Drupal7 $drupal7
-   \*   The Drupal7 service.

File: src/Logger/WatchdogLogger.php [hunk 5]
```diff
@@ -18,6 +19,20 @@
...
+   * The Drupal7 service.
+   *
+   * @var \Drupal\service_container\Legacy\Drupal7
+   */
+  protected $drupal7;
+
+  /**
+   * {@inheritdoc}
+   */
+  function __construct(Drupal7 $drupal7) {
```
Comment (fabianx-ai) [inline src/Logger/WatchdogLogger.php]: Missing **public** function

**Comment by fabianx-ai** [comment]
Little issues, then can go in, taking care of scrutinizer issues and performance in https://www.drupal.org/node/2502517.

**Comment by fabianx-ai** [comment]
Needs rebase after core update

File: src/Legacy/Drupal7.php [hunk 6]
```diff
@@ -11,7 +11,213 @@
...
+   * @param $context3
+   *   (optional) An additional variable that is passed by reference. This
+   *   parameter is deprecated and will not exist in Drupal 8; consequently, it
+   *   should not be used for new Drupal 7 code either. It is here only for
+   *   backwards compatibility with older code that passed additional arguments
+   *   to drupal_alter().
+   */
+  public function drupal_alter($type, &$data, &$context1 = NULL, &$context2 = NULL, &$context3 = NULL) {
+    drupal_alter($type, $data, $context1, $context2, $context3);
+  }
```
Comment (fabianx-ai) [inline src/Legacy/Drupal7.php:350]: Lets also add module_invoke() here.

File: src/Logger/WatchdogLogger.php (hunk 2 above)
Comment (fabianx-ai) [inline src/Logger/WatchdogLogger.php]: @drupol ^^

```

+   * @param \Drupal\service_container\Legacy\Drupal7 $drupal7
+   *   The Drupal7 service.
```

is missing.

File: src/Logger/WatchdogLogger.php (hunk 5 above)
Comment (fabianx-ai) [inline src/Logger/WatchdogLogger.php]: @drupol ^^

File: src/Legacy/Drupal7.php [hunk 7]
```diff
@@ -11,7 +11,213 @@
  * Defines the Drupal 7 legacy service.
  */
 class Drupal7 {
+
+
+  /**
+   * @method void flood_register_event
+   * @method void flood_clear_event
+   * @method bool flood_is_allowed
```
Comment (fabianx-ai) [inline src/Legacy/Drupal7.php]: Per http://www.phpdoc.org/docs/latest/references/phpdoc/tags/method.html

the @method calls need to be on the class.

They also need to have type hinted parameters defined.

Comment (fabianx-ai) [inline src/Legacy/Drupal7.php]: Also lets add all other methods we use here in this PR, so we have 0 new scrutinizer CI issues. :)

File: src/Legacy/Drupal7.php (hunk 6 above)
Comment (drupol) [inline src/Legacy/Drupal7.php:350]: Done.

File: tests/src/StringTranslation/StringTranslationTest.php [hunk 8]
```diff
@@ -15,10 +16,18 @@
...
+   * @var \Drupal\service_container\Legacy\Drupal7
+   */
+  protected $drupal7;
+
+  /**
    * {@inheritdoc}
    */
   public function setUp() {
-    $this->string_translation = new StringTranslation();
+    $this->drupal7 = Mockery::mock('\Drupal\service_container\Legacy\Drupal7');
```
Comment (fabianx-ai) [inline tests/src/StringTranslation/StringTranslationTest.php]: Lets use \Mockery::

File: tests/src/StringTranslation/StringTranslationTest.php [hunk 9]
```diff
@@ -8,6 +8,7 @@
 namespace Drupal\Tests\service_container\StringTranslation;
 
 use Drupal\service_container\StringTranslation\StringTranslation;
+use Mockery;
```
Comment (fabianx-ai) [inline tests/src/StringTranslation/StringTranslationTest.php]: nit - Lets use \Mockery:: and remove the use here

File: tests/src/StringTranslation/StringTranslationTest.php [hunk 10]
```diff
@@ -15,10 +16,18 @@
...
+   */
+  protected $drupal7;
+
+  /**
    * {@inheritdoc}
    */
   public function setUp() {
-    $this->string_translation = new StringTranslation();
+    $this->drupal7 = Mockery::mock('\Drupal\service_container\Legacy\Drupal7');
+    $this->string_translation = new StringTranslation($this->drupal7);
```
Comment (fabianx-ai) [inline tests/src/StringTranslation/StringTranslationTest.php:36]: Lets add a protected variable for StringTranslation

File: src/Legacy/Drupal7.php [hunk 11]
```diff
@@ -9,9 +9,260 @@
 
 /**
  * Defines the Drupal 7 legacy service.
+ *
+ * @method void flood_register_event(string $name, integer $window, string $identifier)
+ * @method void flood_clear_event(string $name, string $identifier)
+ * @method bool flood_is_allowed(string $name, integer $threshold, integer $window, string $identifier)
+ * @method bool module_load_all(bool $bootstrap)
```
Comment (fabianx-ai) [inline src/Legacy/Drupal7.php]: bool $bootstrap = FALSE

etc.

So need to add all the things that are optional here.

**Comment by fabianx-ai** [comment]
Scrutinizer has a bug here:

Opened https://github.com/scrutinizer-ci/php-analyzer/issues/656

This is good to go.

**Pull request merged**